src/analysis/signature_help.ml

```diff
@@ -0,0 +1,137 @@
+open Std
+
+open Browse_raw
+
+open Extend_protocol.Reader
+
+type application_signature =
+  { fun_name : string option
+  ; signature : string
+  ; param_offsets : (int * int) list
+  ; active_param : int option
+  }
+
+(* extract a properly properly parenthesized identifier
+   from (expression_desc (Texp_ident (Longident))) *)
+let extract_ident (exp_desc : Typedtree.expression_desc) =
+  let rec longident ppf : Longident.t -> unit = function
+    | Lident s -> fprintf ppf "%s" (Completion.parenthesize_name s)
+    | Ldot (p, s) -> fprintf ppf "%a.%s" longident p (Completion.parenthesize_name s)
+    | Lapply (p1, p2) -> fprintf ppf "%a(%a)" longident p1 longident p2
+  in
+  match exp_desc with
+  | Texp_ident (_, { txt = li; _ }, _) ->
+    let ppf, to_string = Format.to_string () in
+    longident ppf li;
+    Some (to_string ())
+  | _ -> None
+
+
+(* Type variables shared across arguments should all be
+   printed with the same name.
+   [Printtyp.type_scheme] ensure that a name is unique within a given
+   type, but not across different invocations.
+   [reset] followed by calls to [mark_loops] and [type_sch] provide
+   that *)
+let pp_type env ppf ty =
+  let module Printtyp = Type_utils.Printtyp in
+  Printtyp.wrap_printing_env env ~verbosity:0 (fun () ->
+    Printtyp.mark_loops ty;
+    Printtyp.type_sch ppf ty)
+
+(* surround function types in parentheses *)
+let pp_parameter_type env ppf ty =
+  match ty with
+  | { Types.desc = Tarrow _; _ } as ty ->
+    Format.fprintf ppf "(%a)" (pp_type env) ty
+  | ty -> pp_type env ppf ty
+
+
+(* print parameter labels and types *)
+let pp_parameter env label ppf ty =
+  match label with
+  | Asttypes.Nolabel ->
+    pp_parameter_type env ppf ty
+  | Asttypes.Labelled l ->
+    Format.fprintf ppf "%s:%a" l (pp_parameter_type env) ty
+  | Asttypes.Optional l ->
+    (* unwrap option for optional labels the same way as
+       [Raw_compat.labels_of_application] *)
+    let unwrap_option ty = match (Ctype.repr ty).Types.desc with
+      | Types.Tconstr (path, [ty], _)
+        when Path.same path Predef.path_option -> ty
+      | _ -> ty
+    in
+    Format.fprintf ppf "?%s:%a" l (pp_parameter_type env) (unwrap_option ty)
+
+(* record buffer offsets to be able to underline parameter types *)
+let print_parameter_offset buffer env label ty =
+  let param_start = Buffer.length buffer in
+  Format.fprintf (Format.formatter_of_buffer buffer)
+    "%a%!" (pp_parameter env label) ty;
+  let param_end = Buffer.length buffer in
+  Buffer.add_string buffer " -> ";
```

Comment:
You might as well use `Format.pp_print_string` here. Not a big deal, but it will be more robust if we use boxes.